### DISCUSSION TITLE
DAI 2023-2024 - Check out my application protocol! - Pirakas Anthon

### DISCUSSION THREAD

**Discussion by apirakas** [heig-vd-dai-course/heig-vd-dai-course#126] [Show and tell]
#### Better understand the SMTP protocol

- **What is the purpose of the SMTP protocol?**

  The Simple Mail Transfer Protocol (SMTP) is a protocol used for sending and receiving email messages

- **On which port does the SMTP protocol work?**

  SMTP servers commonly use the Transmission Control Protocol on port number 25 (for plaintext) and 465, 587 (for encrypted (TLS/STARTTLS) communications)

- **On which protocol does the SMTP protocol work?**

  Typically a TCP connection

- **Who initiates the connection?**

  Once the transmission channel is established and initial handshaking is completed, the SMTP client normally initiates a mail transaction.

- **What are the available messages?**

  - HELO/EHLO: initiates the start of the SMTP session and identifies the sender-SMTP to the receiver-SMTP

  - MAIL FROM: specifies the email address of the sender

  - RCPT TO: specifies the email address of the recipient

  - DATA: initiates the transfer of the message content

  - QUIT: ends the SMTP session

    

#### Better understand the POP3 protocol

- **What is the purpose of the POP3 protocol?**

  The Post Office Protocol version 3 (POP3) is used by local email software clients to retrieve emails from a remote mail server.

- **On which port does the POP3 protocol work?**

  The Post Office Protocol version 3 (POP3) works on the following two ports by default
  Port 110: default, non-encrypted port
  Port 995: should be used when the user needs to connect using POP3 securely

- **On which protocol does the POP3 protocol work?**

  Typically a TCP connection

- **Who initiates the connection?**

  The client initiates the connection by connecting to the POP3 server. Once the connection is established, the POP3 server sends a greeting, and the session enters the authorization state. The client then provides login credentials for authentication, and upon successful authentication, a session is initialized, allowing the client to interact with the mailbox

- **What are the available messages?**

  - USER: specifies the mailbox username
  - PASS: specifies the mailbox password
  - STAT: returns the number and total size of all messages
  - LIST: returns the message number and size of each message
  - RETR: retrieves a selected message
  - DELE: deletes a selected message
  - RSET: resets the mailbox, undeleting deleted messages
  - TOP: retrieves the headers and a specified number of lines of a selected message
  - QUIT: ends the POP3 session

- **What is the difference between POP3 and SMTP?**

  SMTP is used for sending emails, while POP3 is used for accessing messages. SMTP is a push protocol, while POP3 is a pop protocol. SMTP sends the email from the sender's device to the receiver's mailbox, and POP3 retrieves and organizes emails from the receiver's mail server to the receiver's computer.

  

#### Better understand the IMAP protocol

- **What is the purpose of the IMAP protocol?**

  IMAP (Internet Message Access Protocol) is an email retrieval protocol that allows users to access their email messages from a remote mail server and manipulate them as if they were stored locally on their device.

- **On which port does the IMAP protocol work?**

  The IMAP (Internet Message Access Protocol) protocol works on two ports by default :

  Port 143: the default port for IMAP, which does not provide any encryption
  Port 993: the secure port for IMAP, which works over TLS/SSL encryption

- **On which protocol does the IMAP protocol work?**

  Typically a TCP connection

- **Who initiates the connection?**

  The client initiates the connection by sending a request to the IMAP server on the specified port number. Once the connection is established, the client sends commands to the server, which replies with responses and/or email message contents

- **What are the available messages?**

  - SELECT: selects a mailbox for subsequent message searching, retrieval, and manipulation
  - EXAMINE: selects a mailbox for subsequent message searching and retrieval, but not for message manipulation
  - CREATE: creates a new mailbox with the specified name
  - DELETE: deletes the mailbox with the specified name
  - RENAME: renames the mailbox with the specified name
  - SUBSCRIBE: adds the specified mailbox name to the server's set of "active" or "subscribed" mailboxes
  - UNSUBSCRIBE: removes the specified mailbox name from the server's set of "active" or "subscribed" mailboxes
  - LIST: returns a list of mailboxes that match the specified pattern
  - STATUS: returns the status of the specified mailbox
  - FETCH: retrieves the specified message(s) from the mailbox
  - STORE: alters the flags associated with the specified message(s)
  - SEARCH: searches the mailbox for messages that match the specified criteria
  - UID: prefixes a command with the unique identifier (UID) of a message, rather than its sequence number

- **What is the difference between POP3, SMTP and IMAP?**

  SMTP is used for sending emails, while POP3 and IMAP are used for accessing messages. SMTP is a push protocol, while POP3 and IMAP are client-server protocols. SMTP sends the email from the sender's device to the receiver's mailbox, and POP3 retrieves and organizes emails from the receiver's mail server to the receiver's computer. IMAP allows users to access their email messages from multiple devices and locations, while keeping the messages stored on the server, and it allows users to organize their email messages into folders and subfolders, and it supports advanced search capabilities.

#### Better understand the SSH protocol

- **What is the purpose of the SSH protocol?**

  The purpose of the SSH (Secure Shell) protocol is to provide a secure way for users to access, control, and modify their remote servers over the internet

- **On which port does the SSH protocol work?**

  The SSH protocol works on port number 22 by default.

- **On which protocol does the SSH protocol work?**

  Operates on top of the Transmission Control Protocol (TCP) for communication

- **Who initiates the connection?**

  In the SSH protocol, the client initiates the connection by sending a request to the SSH server on the specified port number. Once the connection is established, the client sends commands to the server, which replies with responses and/or executes the commands

- **What are the available messages?**

  - SSH_MSG_DISCONNECT: disconnects the client from the server
  - SSH_MSG_IGNORE: ignores the message
  - SSH_MSG_UNIMPLEMENTED: indicates that the requested message is not implemented
  - SSH_MSG_DEBUG: sends a debug message to the client
  - SSH_MSG_SERVICE_REQUEST: requests a service from the server
  - SSH_MSG_SERVICE_ACCEPT: accepts a service request from the client
  - SSH_MSG_KEXINIT: initiates the key exchange process
  - SSH_MSG_NEWKEYS: indicates that new keys will be used for encryption
  - SSH_MSG_USERAUTH_REQUEST: requests user authentication
  - SSH_MSG_USERAUTH_FAILURE: indicates that user authentication has failed
  - SSH_MSG_USERAUTH_SUCCESS: indicates that user authentication has succeeded
  - SSH_MSG_CHANNEL_OPEN: opens a new channel for communication
  - SSH_MSG_CHANNEL_OPEN_CONFIRMATION: confirms the opening of a new channel
  - SSH_MSG_CHANNEL_OPEN_FAILURE: indicates that the opening of a new channel has failed
  - SSH_MSG_CHANNEL_DATA: sends data over an open channel
  - SSH_MSG_CHANNEL_EOF: indicates that no more data will be sent over an open channel
  - SSH_MSG_CHANNEL_CLOSE: closes an open channel

- **How is the version exchange done?**

  The SSH version exchange is done during the initial connection setup phase. The SSH client sends a version string to the SSH server, which replies with its own version string. The client and server then negotiate the version of the protocol to be used for the session

#### Better understand the HTTP protocol

- **What is the purpose of the HTTP protocol?**

  The purpose of the HTTP (Hypertext Transfer Protocol) protocol is to enable communication between web servers and clients. It is a request-response protocol that allows clients to request resources, such as web pages and images, from web servers, and receive responses containing the requested resources.

- **On which port does the HTTP protocol work?**

  The HTTP protocol works on port number 80 by default for unsecured connections, and port number 443 for secured connections using HTTPS (HTTP over SSL/TLS).

- **On which protocol does the HTTP protocol work? Is there a difference between HTTP/2 and HTTP/3?**

  The HTTP protocol  operates on top of the Transmission Control Protocol (TCP) for communication.
  HTTP/2 and HTTP/3 are both newer versions of the HTTP protocol, with HTTP/2 using TCP as the underlying transport protocol, and HTTP/3 using QUIC (Quick UDP Internet Connections) as the transport protocol.

- **Who initiates the connection?**

  In the HTTP protocol, the client initiates the connection by sending a request to the server on the specified port number. Once the connection is established, the client sends requests to the server, which replies with responses containing the requested resources

- **What are the available messages?**

  - GET: requests a representation of the specified resource
  - POST: submits an entity to the specified resource, often causing a change in state or side effects on the server
  - PUT: replaces the specified resource with the request payload
  - DELETE: deletes the specified resource
  - HEAD: requests a response identical to a GET request, but without the response body
  - OPTIONS: describes the communication options for the target resource
  - CONNECT: establishes a network connection to a resource
  - TRACE: performs a message loop-back test along the path to the target resource

#### The context

You are working for a company that wants to create a new application protocol for their new application. This application will be used to **send text messages between users.**

The users use a **client to interact with the server.**

The server runs on a **specific port (1234).** **The clients connect to the server**. The server can only **accept a limited number of connections (5). The server** can r**efuse a connection if it is already full**.

**Once connected,** the users can **send text messages to other users.**

The **server accepts text messages from the client** and **sends them to the recipient.**

The **server can only accept text messages from users that are connected.** The text **messages can only be of a certain length (100 characters).**

- **What is the purpose of the protocol?**

  The purpose of this protocol is to exchange of text messages between users through a client-server architecture. It enables users to connect to a central server, send text messages to other connected users, and receive messages from others.

- **On which port(s) does the protocol work?**

  The protocol operates on a specific port, which in this case is port 1234.

- **On which protocol(s) does the protocol work?**

  As we need data integrity,TCP provides reliable, connection-oriented communication, which is suitable for text messages.

- **Who initiates the connection?**

  Clients initiate the connection to the server. The clients connect to the server running on port 1234.

- **What are the available messages/actions?**

  - Connect: Clients initiate a connection to the server.
  - Send Message: Connected users can send text messages to other connected users.
  - Receive Message: Users can receive messages sent to them by others.
  - Close : ends the session

- **What is the format of the messages/actions?**

  - Connect: The client initiates a connection to the server on port 1234 using standard TCP socket connection.
  - Send Message: Messages sent by clients must adhere to a predefined format, which could include:
    - Sender's identification (e.g., username)
    - Recipient's identification (e.g., username)
    - Message content (up to 100 characters)
  - Receive Message: Messages received will also have a format that includes the sender's identification and the message content.
  - Close : When the client send his message he will wait for a response if the other client didn't respond on a defined timespan the connection will be closed.

- **Are there any edge cases or error cases? What happens in these cases?**

  - Connection Refused: If the server has reached its limit of 5 concurrent connections, it may refuse any additional connection attempts until a slot becomes available. In this case, the client attempting to connect should be informed that the server is currently full.
  - Message Length Limit: If a user attempts to send a message that exceeds the maximum allowed length (100 characters), the server should reject the message and inform the sender.
  - Unrecognized User or Recipient: If a client tries to send a message to a user who is not connected or if the recipient's username is unrecognized, the server should handle this as an error and inform the sender.
 
### Application protocol using a sequence diagram.

![sequenceDiagram](https://github.com/heig-vd-dai-course/heig-vd-dai-course/assets/71434215/8f9f06ef-485d-4ed6-8aa6-3248b9e7a777)

**Comment by ludelafo** [comment]
Great!

I would make usage of some constants/enums to store the messages/commands/actions as well as the errors. This will help to find the similarities from the description, the schema and the code.
